File: wallet-web/app/src/config/network.ts
// Runtime testnet/mainnet switch. wallet-backend is deployed once per
// network (they're separate services with separate databases, not one
// service serving two chains - see wallet-backend/DEPLOYMENT.md), so
// "switching network" here really means "point this app at a different
// wallet-backend deployment" - the API base URL and chain ID both change
// together, never independently.
//
// VITE_WALLET_BACKEND_URL_TESTNET/_MAINNET and VITE_CHAIN_ID_TESTNET/
// _MAINNET (see .env.example) are build-time Vite values, baked into the
// bundle like any other VITE_* var - both networks' values ship in the
// same build, and the choice of which one is active is a runtime toggle
// stored in localStorage, not a rebuild. This lets one deployed app
// (e.g. a staging build) offer the in-app switch this file exists for;
// see PLAN.md and DEPLOYMENT.md's CSP note for why both origins must be
// allowlisted at build time either way.
export type NetworkEnv = 'mainnet' | 'testnet';

export interface NetworkConfig {
  env: NetworkEnv;
  label: string;
  backendUrl: string;
  chainId: number;
}

const STORAGE_KEY = 'trovo.activeNetwork';

function readEnv(name: string, fallback: string): string {
  return (import.meta.env as unknown as Record<string, string | undefined>)[name] ?? fallback;
}

const NETWORKS: Record<NetworkEnv, NetworkConfig> = {
  testnet: {
    env: 'testnet',
    label: 'Base Sepolia (Testnet)',
    backendUrl: readEnv('VITE_WALLET_BACKEND_URL_TESTNET', 'http://localhost:8080'),
    chainId: Number(readEnv('VITE_CHAIN_ID_TESTNET', '84532')),
  },
  mainnet: {
    env: 'mainnet',
    label: 'Base Mainnet',
    backendUrl: readEnv('VITE_WALLET_BACKEND_URL_MAINNET', ''),
    chainId: Number(readEnv('VITE_CHAIN_ID_MAINNET', '8453')),
  },
};

// A build that only ever targets one network (the common case - most
// deployments are one app build per environment, per DEPLOYMENT.md) can
// leave VITE_DEFAULT_NETWORK unset; it defaults to testnet for the same
// forgotten-env-var-safety reason wallet-backend's own BASE_CHAIN_ID
// defaults to Sepolia.
const DEFAULT_NETWORK: NetworkEnv = readEnv('VITE_DEFAULT_NETWORK', 'testnet') === 'mainnet' ? 'mainnet' : 'testnet';

export function getActiveNetwork(): NetworkEnv {
  try {
    const stored = window.localStorage.getItem(STORAGE_KEY);
    if (stored === 'mainnet' || stored === 'testnet') return stored;
  } catch {
    // localStorage unavailable (private browsing, etc.) - fall back to the build default below.
  }
  return DEFAULT_NETWORK;
}

export function getNetworkConfig(network: NetworkEnv = getActiveNetwork()): NetworkConfig {
  return NETWORKS[network];
}

export function isNetworkConfigured(network: NetworkEnv): boolean {
  return NETWORKS[network].backendUrl !== '';
}

/**
 * Persists the chosen network and reloads the app. A reload (rather than
 * an in-place state update) is deliberate: testnet and mainnet are
 * different wallet-backend deployments with entirely separate user
 * registrations, wallet addresses, and balances - there is no
 * "translating" in-memory Redux/cache state from one to the other, so
 * the safest thing is to let the app re-initialize from scratch exactly
 * as it would on a fresh visit, now pointed at the new backend.
 */
export function switchActiveNetwork(network: NetworkEnv): void {
  try {
    window.localStorage.setItem(STORAGE_KEY, network);
  } catch {
    // If we can't persist it, the reload below would just come back to
    // the build default - nothing further we can do without storage.
  }
  window.location.reload();
}

File: wallet-mobile/app/lib/config/network.dart
// Runtime testnet/mainnet switch - the mobile counterpart to wallet-web's
// config/network.ts. Same reasoning: wallet-backend is deployed once per
// network (separate services, separate databases), so "switching
// network" means pointing this app at a different wallet-backend
// deployment, never translating in-memory state between them.
//
// Vite's build-time VITE_* env vars become Dart's compile-time
// `--dart-define` values (String.fromEnvironment) - baked into the build
// the same way, read at compile time rather than runtime. The active
// choice itself is a runtime toggle in SharedPreferences, matching
// wallet-web's localStorage.
import 'package:shared_preferences/shared_preferences.dart';

enum NetworkEnv { testnet, mainnet }

class NetworkConfig {
  const NetworkConfig({required this.env, required this.label, required this.backendUrl, required this.chainId});

  final NetworkEnv env;
  final String label;
  final String backendUrl;
  final int chainId;

  bool get isConfigured => backendUrl.isNotEmpty;
}

const _storageKey = 'trovo.activeNetwork';

const _networks = <NetworkEnv, NetworkConfig>{
  NetworkEnv.testnet: NetworkConfig(
    env: NetworkEnv.testnet,
    label: 'Base Sepolia (Testnet)',
    backendUrl: String.fromEnvironment('WALLET_BACKEND_URL_TESTNET', defaultValue: 'http://localhost:8080'),
    chainId: int.fromEnvironment('CHAIN_ID_TESTNET', defaultValue: 84532),
  ),
  NetworkEnv.mainnet: NetworkConfig(
    env: NetworkEnv.mainnet,
    label: 'Base Mainnet',
    backendUrl: String.fromEnvironment('WALLET_BACKEND_URL_MAINNET', defaultValue: ''),
    chainId: int.fromEnvironment('CHAIN_ID_MAINNET', defaultValue: 8453),
  ),
};

const _defaultNetwork =
    String.fromEnvironment('DEFAULT_NETWORK', defaultValue: 'testnet') == 'mainnet' ? NetworkEnv.mainnet : NetworkEnv.testnet;

class NetworkSettings {
  NetworkSettings(this._prefs);

  final SharedPreferences _prefs;

  static Future<NetworkSettings> load() async => NetworkSettings(await SharedPreferences.getInstance());

  NetworkEnv getActiveNetwork() {
    final stored = _prefs.getString(_storageKey);
    if (stored == 'mainnet') return NetworkEnv.mainnet;
    if (stored == 'testnet') return NetworkEnv.testnet;
    return _defaultNetwork;
  }

  NetworkConfig getNetworkConfig([NetworkEnv? network]) => _networks[network ?? getActiveNetwork()]!;

  bool isNetworkConfigured(NetworkEnv network) => _networks[network]!.isConfigured;

  /// Persists the chosen network. Unlike wallet-web (which reloads the
  /// whole page - the only way a browser tab can safely re-initialize
  /// from scratch), the caller here is expected to pop back to a fresh
  /// root widget itself (e.g. re-run the app's root MaterialApp) after
  /// this returns, for the same reason: testnet and mainnet are separate
  /// backends with nothing in memory worth carrying across.
  Future<void> switchActiveNetwork(NetworkEnv network) async {
    await _prefs.setString(_storageKey, network.name);
  }
}
